Site Editor: Add new Comments Query Loop block

michalczaplinski opened the pull request:
## Description
Fixes https://github.com/WordPress/gutenberg/issues/34996
Fixes https://github.com/WordPress/gutenberg/issues/35000

- This PR is adding a new Comments Query block to `block-library`.
- It includes only the bare skeleton of the block without all the attributes that should be present in the final version.
- I've included a stub implementation of the [`core/comment-template` block](https://github.com/WordPress/gutenberg/issues/35000) which is kept to the bare minimum to make the comments query work.

## How has this been tested?
Tested manually:

1. Create a post
2. Type `/comment`
3. Select the `Comment Query Loop` block
4. Change the `perPage`, `pages` or `offset` attributes in the toolbar and observe them change in the UI

### Environment details
- WordPress Version: 5.9-alpha-51831
- Database: `wp-env` default
- PHP Version: `wp-env` default
- Node: `14.17.6`
- npm: `6.14.15`

## Screenshots
<img width="600" alt="Screen Shot 2021-10-07 at 20 09 01" src="https://user-images.githubusercontent.com/5417266/136482656-650c5a47-59a8-4d72-94ad-e7f5fbb55b3a.png">
<img width="600" alt="Screen Shot 2021-10-07 at 20 09 35" src="https://user-images.githubusercontent.com/5417266/136482697-fba25b2e-0b0a-4e53-9b78-be83e98fbf1a.png">
<img width="600" alt="Screen Shot 2021-10-07 at 20 10 57" src="https://user-images.githubusercontent.com/5417266/136482791-45bb94d9-4795-4c4d-bd3f-85a0133e38de.png">
<img width="619" alt="Screen Shot 2021-10-07 at 20 13 45" src="https://user-images.githubusercontent.com/5417266/136482972-8a6f2f96-faa3-440e-829f-4d1fdbe3e949.png">

## Types of changes
- Type: Experimental
- Needs Testing
- New feature (Block Editor)

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

michalczaplinski committed:
Add most basic skeleton for Comments Query Loop block

michalczaplinski committed:
Add a stub for core/comments-template block

michalczaplinski committed:
Clean up the implementation of core/comments-query

michalczaplinski committed:
Move query-content & query-setup to separate files

michalczaplinski committed:
Remove the post-comments-query block

michalczaplinski committed:
Refactor getFirstQueryClientIdFromBlocks to utils.js

michalczaplinski committed:
Add the variations of the comments query block

michalczaplinski committed:
Remove unnecesary parameters from block.json

michalczaplinski committed:
Add a proper query-placeholder component

michalczaplinski committed:
Add instruction about where to add new integration test fixtures

cbravobernal commented on packages/block-library/src/comments-template/block.json:
```diff
@@ -0,0 +1,26 @@
+	"title": "Comments Template",
+	"category": "design",
+	"parent": [ "core/comments-query" ],
+	"description": "Contains the block elements used to render a comment, like the title, date, author, avator and more.",
```
avator -> typo alert!

michalczaplinski replied:
I like "avator" better! 😂

michalczaplinski committed:
Fix typo in comments-template block.json

michalczaplinski committed:
Add extra space in error message in full-content.test.js

michalczaplinski committed:
Add integration test fixtures for new blocks

michalczaplinski committed:
Add postComments icon to core/comments-query block

michalczaplinski committed:
Add offset back to comments-query block

gziolo commented on packages/block-library/src/comments-query/block.json:
```diff
@@ -0,0 +1,45 @@
+		"queryId": {
+			"type": "number"
+		},
+		"query": {
```
@ntsekouras, since it's a new block, do you still think that using an object for all query-related attributes was a good design choice? Here, we could promote all individual fields to become block attributes:
- `queryPerPage`
- `queryTotalPages`
- `queryOrder`
- `queryOrderBy`

The benefit would be that you would explicitly define a type of all those fields, and the default value wouldn't have to be overridden for all fields when only one of them needs to change when applying to the block variation. Everything else would remain more or less the same. In the Comment Template block, you could access the context almost the same way: `queryPerPage` vs `query.perPage`.

ntsekouras replied:
>do you still think that using an object for all query-related attributes was a good design choice? Here, we could promote all individual fields to become block attributes:

With `Query Loop` I tried many times to change our `object` attributes handling (like `default` values etc..) that led to the conclusion not to do it, because of the existing behavior for so long that could cause more problems about what people expect with their code.

Having said that, all the points by Grzegorz are spot on and especially if we have only a few attributes, it would be better to have them as top level attributes.

michalczaplinski replied:
Got it - makes sense here. 👍

gziolo commented on packages/block-library/src/comments-query/edit/index.js:
```diff
@@ -0,0 +1,26 @@
+			!! select( blockEditorStore ).getBlocks( clientId ).length,
+		[ clientId ]
+	);
+	const Component = hasInnerBlocks ? QueryContent : QueryPatternSetup;
```
I think you can skip `QueryPatternSetup` for the initial implementation. I don't think we even have any patterns to pick from for now. @ntsekouras should know best how to approach all that.

michalczaplinski replied:
That would mean that we don't need any variations for the Comments Query either, at least not at this stage, correct?

ntsekouras replied:
>Definitely, a good start would be to have a Comments Query Loop and Comment template block with some enforced layout for start

I agree 100%. We just need to have a default `template` for the inner blocks. No variations or patterns are needed for the first iteration. Pinging @kjellr for awareness regarding such patterns.

--edit: I just saw that the comment is outdated 😄

gziolo replied:
Thank you for confirmation Nik 👍🏻

gziolo reviewed (commented):
Query Loop is quite a big block with several functionalities added over time. It took months to develop it. I guess the best approach would be to start simple and code only a small subset of features. @ntsekouras might have the best recommendation here. Definitely, a good start would be to have a Comments Query Loop and Comment template block with some enforced layout for start with simple controls that change some common query attributes.

michalczaplinski committed:
Delete files that are out of scope for this PR

michalczaplinski committed:
Remove variations from comments-query

michalczaplinski committed:
Refactor edit component of comments-query

michalczaplinski committed:
Remove unused utils

michalczaplinski committed:
Comments Query: Refactor edit and toolbar

michalczaplinski committed:
Comments Query: Update values of perPage, page & offset attributes

michalczaplinski commented on packages/block-library/src/comments-query/edit.js:
```diff
@@ -0,0 +1,69 @@
+
+	return (
+		<>
+			{ /* XXX TODO: Add query inspector controls here */ }
```
This will be done in another PR to keep the scope of the current one small.

ntsekouras replied:
If you implement it in a different PR, it would be better to skip the comment 😄

michalczaplinski committed:
Add a CSS class to the QueryToolbar

michalczaplinski committed:
CommentsQuery: Update function name

michalczaplinski committed:
Remove order and orderBy from block.json

michalczaplinski commented on test/integration/full-content/full-content.test.js:
```diff
@@ -266,7 +266,9 @@ describe( 'full post content fixture', () => {
 					errors.push(
 						format(
-							"Expected a fixture file called '%s.html' or '%s__*.html'.",
+							"Expected a fixture file called '%s.html' or '%s__*.html' in `test/integration/fixtures/blocks/` " +
+								'\n\n' +
+								'For more information on how to create test fixtures see https://github.com/WordPress/gutenberg/blob/1f75f8f6f500a20df5b9d6e317b4d72dd5af4ede/test/integration/fixtures/blocks/README.md\n\n',
```
This change is unrelated to the rest of the PR. I've noticed that when you add a new block, the `full-content` integration tests fail and even tell you what file you should create which is 💯 . It was still missing the information about where that file is expected to be and what content to put in there which should now be more clear 🙂. 

I've included this change in this PR instead making a new one because I wanted to make sure that it gives that updated error when you create a new block (which this PR is doing :))

gziolo replied:
It's up to you. You can also open a new PR. @c4rl0sbr4v0 run into exactly the same issue:
https://github.com/WordPress/gutenberg/pull/35396#pullrequestreview-773624216

The most important part here is we improve the message so it's easy to find more details on how to add those fixtures and why.

michalczaplinski replied:
Will keep that change here in that case 🙂

michalczaplinski commented on packages/block-library/src/comments-query/index.js:
```diff
@@ -0,0 +1,20 @@
+/**
+ * WordPress dependencies
+ */
+import { postComments as icon } from '@wordpress/icons';
```
using this icon for now as [suggested](https://github.com/WordPress/gutenberg/issues/34996#issuecomment-931153524) by @jameskoster

michalczaplinski committed:
Comments Template: Stringify the query

michalczaplinski committed:
Update the integration test fixture

michalczaplinski committed:
Update icon for comments-template

michalczaplinski commented on packages/block-library/src/comments-query/toolbar.js:
```diff
@@ -0,0 +1,109 @@
+										min={ 1 }
+										max={ 100 }
+										onChange={ ( value ) => {
+											const num = parseInt( value, 10 );
```
Majority of this file is copied over from `core/query`, but I've added this line in the `onChange` handlers because `value` here is a string and we want to treat it as a number.

Reported this in https://github.com/WordPress/gutenberg/issues/33291#issuecomment-930375758

michalczaplinski committed:
Remove unnecessary CSS

michalczaplinski reviewed (commented):
Thanks for the comments @gziolo! I've reduced the scope of this PR to remove the variations, and `CommentsQuerySetup` and kept just the 3 attributes: `perPage`, `pages` and `offset` for now.

I would appreciate some more eyes on this PR - perhaps @ntsekouras also could take a look to let me know what else you would like to see here in order for this PR to land?

michalczaplinski commented:
No idea why e2e tests are [failing](https://github.com/WordPress/gutenberg/runs/3833729587#step:6:21) but it seems unrelated...? 🤔

michalczaplinski commented:
I think this block should have e2e tests [similar to ones for the `query`](https://github.com/WordPress/gutenberg/blob/trunk/packages/e2e-tests/specs/editor/blocks/query.test.js) block, right?

ntsekouras commented:
Thanks for working on this @michalczaplinski - this is awesome!

I'll take a closer look later to review properly but for now I was just wondering if it makes sense to revisit the design/position of the controls. On one hand it would be good to be consistent between similar blocks, on the other hand if the attributes are fewer, we could move them to the `Inspector controls` maybe? We have a [long standing issue](https://github.com/WordPress/gutenberg/issues/25198) about `Query Loop` controls and very often having the controls in the toolbar has come up as a source of confusion for users.

--cc @jameskoster @jasmussen

ntsekouras commented on packages/block-library/src/comments-query/block.json:
```diff
@@ -0,0 +1,44 @@
+			"type": "string",
+			"default": "div"
+		},
+		"displayLayout": {
```
Let's skip this property for now. This property was added before the `layout` advances and the goal is to deprecate it in `Query Loop` as well and use the layout support (`flow/flex` for now).

ntsekouras commented on packages/block-library/src/comments-query/save.js:
```diff
@@ -0,0 +1,14 @@
+import { InnerBlocks, useBlockProps } from '@wordpress/block-editor';
+
+export default function CommentsQuerySave( {
+	attributes: { tagName: Tag = 'div' },
```
I believe the default `div` is not needed here as it's a new block. It exists in `Query Loop` because the `tagName` attribute was added in a later stage.

ntsekouras commented on packages/block-library/src/comments-query/toolbar.js:
```diff
@@ -0,0 +1,111 @@
+} ) {
+	return (
+		<>
+			{ ! query.inherit && (
```
I guess we will not need `inherit` handling here, will we? 🤔 I'll have to check more about the retrieval of comments(`WP_Comment_Query`), but I think the `inherit` concept doesn't apply here..

michalczaplinski replied:
Oh, I see! I think this was just copy-pasted from the Query Loop block. 

So to help me understand completely:

- A Query Loop block can be inside of a [block template](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-templates/). 
- If that's the case, the block template can pass a `query` attribute to this Query Loop block. 
- If it does so, then it should _also_ pass an `inherit: true` attribute to the Query Loop block [in order to not show the toolbar](https://github.com/WordPress/gutenberg/blob/569e24d8c20baaa5004e73f9573a01e7316d5e43/packages/block-library/src/query/edit/query-inspector-controls.js#L180-L187).

I'm talking about the Query Loop block here but the same mechanism can be applied to other blocks that are inside of a block template. 

Is that correct?

michalczaplinski replied:
So, would we not need `inherit` after all in case that the Comments Query block is inside of a block template? 🤔

ntsekouras replied:
Block templates are just a way of showing some blocks in some initial states. They don't have something to do with the `inherit` functionality.
>A block template is defined as a list of block items. Such blocks can have predefined attributes, placeholder content, and be static or dynamic. Block templates allow specifying a default initial state for an editor session.

`inherit` has to do wether we want the `Query Loop` block to inherit attributes from the global WP_Query. That means if we are in an `archive` page the category shown will be from the global query, so it will work as expected in any `category` page. In single posts/pages we don't `inherit` (as the global query has the single page/post itself) and we have a 'custom' query that we adjust through the block's settings.

From the issues description:
>An advanced block that allows displaying the comments of a specific post based on different parameters and visual configurations. It should be similar to the Query Loop block.

As I understand it, the `comments loop` block is going to always use the provided/current `postId`, no? Will we have a case that we explicitly set `commentIds` or another `postId`? 🤔

michalczaplinski replied:
> inherit has to do wether we want the Query Loop block to inherit attributes from the global WP_Query. That means if we are in an archive page the category shown will be from the global query, so it will work as expected in any category page

ah, gotcha. I now see that it's [used via context ](https://github.com/WordPress/gutenberg/blob/2ecd5a0a373c23e8046193c7f681223f62e20cd3/packages/block-library/src/post-template/index.php#L23)in `post-template/index.php`.

michalczaplinski replied:
> As I understand it, the comments loop block is going to always use the provided/current postId, no? Will we have a case that we explicitly set commentIds or another postId? 🤔

I'm not sure about this one. @SantosGuillamot Do you know the requirements here in more detail? Is the Comments Query block always going to be using the current (default) `postId` or the one provided by the user? Or is there a use case where we might want to select the `commentId(s)` that we want to be shown in this block?

In any case, I think we can leave it as it is now (without using the `inherit`) and implement if needed in a future iteration.

SantosGuillamot replied:
>  @SantosGuillamot Do you know the requirements here in more detail? Is the Comments Query block always going to be using the current (default) postId or the one provided by the user? Or is there a use case where we might want to select the commentId(s) that we want to be shown in this block?

I wouldn't consider it a requirement. I cannot think of a use case right now, so I agree that we can leave it without that functionality right now and, in case it is needed, add it in the future.

michalczaplinski replied:
ok, thanks @SantosGuillamot 🙂 . So, I will leave the `inherit` property out out of this PR for now.

ntsekouras commented on packages/block-library/src/comments-query/toolbar.js:
```diff
@@ -0,0 +1,111 @@
+								</BaseControl>
+								<BaseControl
+									help={ __(
+										'Limit the pages you want to show, even if the query has more results. To show all pages use 0 (zero).'
```
Should we skip this for now as we don't have any Comment Pagination blocks?

michalczaplinski replied:
gotcha!

ntsekouras commented on packages/block-library/src/comments-template/block.json:
```diff
@@ -0,0 +1,24 @@
+		"reusable": false,
+		"html": false,
+		"align": true,
+		"__experimentalLayout": {
```
I think we should check `layouts` better since we have a fresh start for this. We could possibly skip this as well for a first iteration.

ntsekouras commented on packages/block-library/src/comments-template/index.js:
```diff
@@ -0,0 +1,20 @@
+/**
+ * WordPress dependencies
+ */
+import { layout as icon } from '@wordpress/icons';
```
Nitpick: should this icon change?

michalczaplinski replied:
I'm not sure. What do you think @jameskoster? Should I be using the same icon here as for Post Template?

ntsekouras replied:
Or maybe a new one..

ntsekouras commented on packages/block-library/src/index.js:
```diff
@@ -167,6 +169,9 @@ export const __experimentalGetCoreBlocks = () => [
 	// Theme blocks
+	commentsQuery,
+	commentsTemplate,
+
```
an extra line here 😄

ntsekouras commented on packages/block-library/src/comments-query/edit.js:
```diff
@@ -0,0 +1,69 @@
+const TEMPLATE = [ [ 'core/comments-template' ] ];
+
+export default function CommentsQueryEdit( { attributes, setAttributes } ) {
+	const { query, tagName: TagName = 'div', layout = {} } = attributes;
```
Same comment for `tagName`, no need to add a default here.

ntsekouras commented on packages/block-library/src/comments-query/edit.js:
```diff
@@ -0,0 +1,69 @@
+		return { themeSupportsLayout: getSettings()?.supportsLayout };
+	}, [] );
+
+	const defaultLayout = useSetting( 'layout' ) || {};
```
In order to use the `default(flow)` layout we need to add `"__experimentalLayout": true` to this block's `block.json` file - currently is not set. This is also part of deciding how to handle `layout` for these blocks.

michalczaplinski replied:
Hmm, I don't fully understand yet how those layouts are defined. This is referring to the [layout setting of a theme](https://developer.wordpress.org/block-editor/how-to-guides/themes/create-block-theme/#content-width-and-theme-support-for-wide-and-full-width-blocks), right? 

It's not clear to me what options I have for handling the layout here... Could you point me to some examples of blocks that handle layout in different ways?

Also, what is `default(flow)` ? 🙂

ntsekouras replied:
>Also, what is default(flow) ? 🙂

The `flow` layout is an alias for the `default`, which means blocks are placed one below the other (ex. Group block).

In order to define a `layout` for a block you need to have `__experimentalLayout` in its `block.json`. This object can have many settings for now (check below PR with flex-Social Icons, and [this one](https://github.com/WordPress/gutenberg/pull/34194)), but if we just set it to `true` it will use the `default/flow` layout.

In general `layout` can be used by 'container' blocks to declaratively handle how their nested blocks should display. This is something we will probably need here, but can implement in follow ups PRs. If we omit for now the `layout` it means that the `comments loop` will take all the available space given from its parent container. Makes sense? 

Here is the [dev note](https://make.wordpress.org/core/2021/06/29/on-layout-and-content-width-in-wordpress-5-8/) for layout. And here is an example with [flex layout](https://github.com/WordPress/gutenberg/pull/34493/). You can check `Group` block for the `flow/default`.

michalczaplinski replied:
ok, thanks that makes sense! So, we can remove those lines for now:  https://github.com/wordpress/gutenberg/blob/258f10c8d16b3a5d7ad08cd4d4dbf455729dfa37/packages/block-library/src/comments-query/edit.js#L31-L43, correct?

ntsekouras replied:
Yes, but we will still need `useBlockProps` and `useInnerBlocksProps` (without passing the __experimentalLayout prop).

`useBlockProps` in general is what 'marks' something as a block, attaching all the event listeners etc..

michalczaplinski replied:
gotcha, thanks!

ntsekouras commented on packages/block-library/src/comments-query/edit.js:
```diff
@@ -0,0 +1,69 @@
+				/>
+			</BlockControls>
+			<InspectorControls __experimentalGroup="advanced">
+				<SelectControl
```
Not for this PR of course but noting that we should extract this component to be reused.

Ref: [here](https://github.com/WordPress/gutenberg/pull/33976#pullrequestreview-745083553) and  [here](https://github.com/WordPress/gutenberg/pull/35203#issuecomment-929832498)

ntsekouras reviewed (commented):
Thanks for working on this Michal! 

This is probably a chance for us to enforce all the new stuff (like `layout`) and handle better some challenges that we faced with `Query Loop`. In general I think more things to discuss will be exposed when you start the implementation of `Comments Loop` client and server side.

jameskoster commented:
> We have a long standing issue about Query Loop controls and very often having the controls in the toolbar has come up as a source of confusion for users.

I would be in favour of maintaining consistency across the query blocks for now, then addressing the issues of the current UX subsequently and applying any fixes to all blocks at the same time.

michalczaplinski committed:
Comments Query: Remove the unnecessary "div"

michalczaplinski committed:
Comments Query: Remove `displayLayout` from block.json

michalczaplinski committed:
Comments Query: remove _experimentalLayout

michalczaplinski committed:
Comments Query: Remove unused queryContext

michalczaplinski committed:
Comments Query: Remove inherit & pages attribute

michalczaplinski committed:
Comments Query: Refactor the attributes

michalczaplinski committed:
Comments Query: Update the offset label

michalczaplinski committed:
Comments Query: Simplify passing of setAttributes

gziolo commented:
We need to sort out the execution plan before we land this PR. At the moment there are two existing blocks:
- `core/post-comments`
- `core/post-comment`

They are going to be replaced with:
- 'core/comments-query'
- core/comments-template'

Now the question is how we approach that:
1. Do we keep both versions of blocks for some time? It would require a manual update for all existing block themes that use `core/post-comments` or `core/post-comment`.
2. Do we keep old (internal) block names and replace them with the revised implementation? Is that even possible from a backward compatibility standpoint?

Do we have other options available that I missed? Is it something that you already know how to address? /cc @mtias and @youknowriad.

youknowriad commented:
Potentially, we could keep the old block only in the plugin and only backport one of them to Core. For Post, we did keep both blocks (latest posts and Query Loop).

gziolo commented:
> Potentially, we could keep the old block only in the plugin and only backport one of them to Core. For Post, we did keep both blocks (latest posts and Query Loop).

I forgot about Latest Comments. It is there as well:

https://github.com/WordPress/gutenberg/blob/39be23fc2fc58b93ddeaf77e7110c87eb3313837/packages/block-library/src/latest-comments/block.json#L3

At the moment, we have the experimental Post Comments block that is very simple and seems to be a blocky version of `comments_template()` call in PHP:

https://github.com/WordPress/gutenberg/blob/39be23fc2fc58b93ddeaf77e7110c87eb3313837/packages/block-library/src/post-comments/index.php#L33


There is also an existing experimental Post Comment block that requires providing the comment id by the user in the editor:

https://github.com/WordPress/gutenberg/blob/39be23fc2fc58b93ddeaf77e7110c87eb3313837/packages/block-library/src/post-comment/edit.js#L42-L47

This PR proposes a Comments Query Loop block that is a more advanced version of Post Comments. The other block Comments Template is very similar to the existing Post Comment block.

michalczaplinski committed:
Comments Query: update the snapshot

michalczaplinski commented:
I'd like to understand what else would be needed in order to land this PR? I only have this on my todo list but there's probably more that I'm missing:

- If confirmed, remove the handling of layout as suggested by Nik in https://github.com/WordPress/gutenberg/pull/35231#discussion_r731039698
- Add E2E tests [like the ones for Query Loop block ](https://github.com/WordPress/gutenberg/blob/trunk/packages/e2e-tests/specs/editor/blocks/query.test.js).

@gziolo @ntsekouras would you like to give it another look?

gziolo commented on lib/blocks.php:
```diff
@@ -77,6 +77,8 @@ function gutenberg_reregister_core_block_types() {
 				'post-comments-form.php'        => 'core/post-comments-form',
 				'post-comments-link.php'        => 'core/post-comments-link',
+				'comments-query.php'            => 'core/comments-query',
+				'comments-template.php'         => 'core/comments-template',
```
We discussed the naming with @ntsekouras and we arrived at the following to match with the Query Loop block:
```suggestion
				'comments-query-loop.php'    => 'core/comments-query-loop',
				'comment-template.php'          => 'core/comment-template',
```

Sidenote: in the code there is `core/query` that has the title Query Loop. It was too late to rename it to `core/query-loop` that would work best.

gziolo replied:
By the way, there is no PHP file for the Comment Template block. We should add one with a basic handling for the comments query loop.

gziolo commented on packages/block-library/src/comments-query/block.json:
```diff
@@ -0,0 +1,39 @@
+	"providesContext": {
+		"queryId": "queryId",
+		"queryPerPage": "queryPerPage",
+		"queryOffset": "queryOffset",
```
We agreed with @ntsekouras that this might cause some conflicts with the context exposed from the Query Loop block, in particular `queryId`. Let's discuss whether we should rename them here to mitigate the risk. In particular, we were talking about displaying a list of posts on the page, that also include a list of comment. A good example would be https://make.wordpress.org/core/

<img width="798" alt="Screen Shot 2021-10-21 at 11 30 56" src="https://user-images.githubusercontent.com/699132/138250957-64c3ab3e-a8da-4aff-b1eb-873ecd5466be.png">

michalczaplinski replied:
1. We could encapsulate everything in an object called `commentsQuery`:

```json
{
	"attributes": {
		"commentsQuery": {
			"queryId": {
				"type": "number"
			},
			"commentsNumber": {
				"type": "number",
				"default": 4
			},
			"offset": {
				"type": "number",
				"default": 0
			}
		}
	},
	"providesContext": {
		"commentsQuery": "commentsQuery"
	}
}
``` 

2. We can rename `queryId` to `commentsQueryId` and then _also_ provide an object for all the query related stuff: 

```json
{
	"attributes": {
		"commentsQueryId": {
			"type": "number"
		},
		"commentsQuery": {
			"commentsNumber": {
				"type": "number",
				"default": 4
			},
			"offset": {
				"type": "number",
				"default": 0
			}
		}
	},
	"providesContext": {
		"commentsQueryId": "commentsQueryId",
    	"commentsQuery": "commentsQuery"
	}
}
```

3. Rename `queryId` to `commentsQueryId` and then pass all the other attributes with a `comments*` prefix:

```json
{
	"attributes": {
		"commentsQueryId": {
			"type": "number"
		},
		"commentsNumber": {
			"type": "number",
			"default": 4
		},
		"commentsOffset": {
			"type": "number",
			"default": 0
		}
	},
	"providesContext": {
		"commentsQueryId": "commentsQueryId",
		"commentsNumber": "commentsNumber",
		"commentsOffset": "commentsOffset"
	}
}
```

It's a bit hard for me to say which one would be most consistent without all that much experience with other blocks. I'm leaning towards **3.** but there's not a clear winner here I think. Do you have any other ideas?

ntsekouras replied:
I guess `option 3` is the one we already discussed above, as it's far easier to work with top level attributes and also the number of the attributes will be far less than `Query Loop` (I think 😄)

gziolo replied:
I checked the docs for [providing block contex](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-context/#providing-block-context)t and it looks like we can alias attribute to resolve the issue with naming conflicts.

Would a slightly modified `option 3` work for you better?

```json
{
	"attributes": {
		"queryId": {
			"type": "number"
		},
		"number": {
			"type": "number",
			"default": 4
		},
		"offset": {
			"type": "number",
			"default": 0
		}
	},
	"providesContext": {
		"commentsQueryId": "queryId",
		"commentsNumber": "number",
		"commentsOffset": "offset"
	}
}
```

ntsekouras replied:
That sounds great Grzegorz! I had forgotten about that 😄

michalczaplinski replied:
> I checked the docs for providing block context and it looks like we can alias attribute to resolve the issue with naming conflicts.

noice, will update the PR to use this :)

ntsekouras commented:
>I'd like to understand what else would be needed in order to land this PR? I only have this on my todo list but there's probably more that I'm missing:

IMO  a first iteration would be to have a block that can be inserted and display the last comments of the current post (using the `postId` provided by the context) and just have a `commentsNumber` as a setting. So I guess it's okay to remove the `offset` as well etc..

What this means is implement the client and server side of the `comments template` block to retrieve some comments for the current post/page displayed and show them without worrying for other nuances like pagination or hierarchical comments etc..

Also since lot's of people working on these comment blocks and quick follow up are about to happen, I'd be okay without adding e2e tests right now, in order to land sooner.

jameskoster commented:
> IMO a first iteration would be to have a block that can be inserted and display the last comments of the current post (using the postId provided by the context) and just have a commentsNumber as a setting. So I guess it's okay to remove the offset as well etc..

It may just be me, but this seems like a fairly odd use case that we probably shouldn't encourage. Why would someone insert the comments _in_ the post? Assuming the comments loop is also placed in the post template, they would be duplicated on the frontend. Also, if you put the comments loop in the post, but not the comment form – how would an end user add comments?

I also don't really see an argument to allow the comments loop to display comments from an arbitrary post. Again, I'm struggling to come up with a scenario where you'd want to display the comments from one post or page in another.

Apologies if I'm missing something obvious, but doesn't this feel like a distinctly template-centric block?

ntsekouras commented:
>It may just be me, but this seems like a fairly odd use case that we probably shouldn't encourage. Why would someone insert the comments in the post? Assuming the comments loop is also placed in the post template, they would be duplicated on the frontend.

I agree with this and I guess this will be used mostly in templates, but the implementation of the block doesn't really care if it lives in a post or a `single post` template. It will still get the `current post id` from the context. We are not going to write any extra code for this - it's just how it works and the same goes with other Post blocks.

>I also don't really see an argument to allow the comments loop to display comments from an arbitrary post. Again, I'm struggling to come up with a scenario where you'd want to display the comments from one post or page in another.

Totally agree with this. Additionally we came to a similar conclusion with @gziolo, whether the current `PostComment` block should be preserved or not, which handles a use case of having to show a specific comment anywhere by its `id`. If we don't find a use case for this, we could remove/deprecate it and only have the `comments template block`(similar to `PostTemplate`) that is something like a 'wrapper' to the existing `PostComment` block.

jameskoster commented:
Understood. I wonder if there isn't a more practical use case we could aim for though? Maybe this is what you meant and I misinterpreted, but perhaps this might be better:

> A first iteration would be to have a block that can be inserted **while editing the template of a post or page** or page and display the last comments of the current post (using the postId provided by the context)

Then we can ignore things like commentsnumber and offset for now and just inherit from whatever is set in the discussion settings.

michalczaplinski committed:
Update lib/blocks.php

Update the naming

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

michalczaplinski committed:
Comments Query rename:
- comments-template -> comment-template
- comments-query -> comments-query-loop

michalczaplinski commented:
> IMO a first iteration would be to have a block that can be inserted and display the last comments of the current post (using the postId provided by the context) and just have a commentsNumber as a setting. So I guess it's okay to remove the offset as well etc..

All righty. I will work on it in this PR here in this case. Inside of the Comment Template I will use the [Comment Date](https://github.com/WordPress/gutenberg/pull/35112), [Comment Content](https://github.com/WordPress/gutenberg/pull/35183) and [Comment Author](https://github.com/WordPress/gutenberg/pull/35595) blocks that have been merged recently. 

> Then we can ignore things like commentsnumber and offset for now and just inherit from whatever is set in the discussion settings.

👍

gziolo commented:
> > I also don't really see an argument to allow the comments loop to display comments from an arbitrary post. Again, I'm struggling to come up with a scenario where you'd want to display the comments from one post or page in another.
>
> Totally agree with this. Additionally we came to a similar conclusion with @gziolo, whether the current PostComment block should be preserved or not, which handles a use case of having to show a specific comment anywhere by its id. If we don't find a use case for this, we could remove/deprecate it and only have the comments template block(similar to PostTemplate) that is something like a 'wrapper' to the existing PostComment block.

I can confirm that. The way the Post Comment (`core/post-comment`) block woks at the moment is very difficult to use and I can't think of a any compelling reason to keep it at the moment. It seems like it should get replaced with the Comment Template (`core/comment-template`) block that @michalczaplinski works on in this PR. As discussed above, we can keep both blocks for some time and decide later what to do about the older one.

We arrived at similar conclusions in regards to the Post Comments (`core/post-comments`) block. We should plan to deprecate it when the Comments Query Loop (`core/comments-query-loop`) is fully functional.

> IMO a first iteration would be to have a block that can be inserted and display the last comments of the current post (using the postId provided by the context) and just have a commentsNumber as a setting. So I guess it's okay to remove the offset as well etc..
>
> What this means is implement the client and server side of the comments template block to retrieve some comments for the current post/page displayed and show them without worrying for other nuances like pagination or hierarchical comments etc..

That sounds like a great plan 💯 We can even harcode the number of comments for start if that causes any implementation issues. I also wanted to share a related code from the Query Loop block that can be used as a reference:

1. Editor:
https://github.com/WordPress/gutenberg/blob/070deca2f5bd7b08da4d35c8f1aa68fa018f625b/packages/block-library/src/post-template/edit.js#L128-L167
2. Frontend:
https://github.com/WordPress/gutenberg/blob/070deca2f5bd7b08da4d35c8f1aa68fa018f625b/packages/block-library/src/post-template/index.php#L53-L74

michalczaplinski committed:
Comment Template block: Add index.php

michalczaplinski committed:
Comments Query: Remove offset attribute

michalczaplinski committed:
Comments Query: Remove outdated fixtures

michalczaplinski committed:
Comments Query: Add @wordpress/editor to deps

michalczaplinski committed:
Comments Query: Add displaying basic comments content

michalczaplinski committed:
Comments Query: Remove more old fixtures

michalczaplinski committed:
Comments Query: Add a very basic server-side render

michalczaplinski committed:
Comments Query: Simplify query for more comments

michalczaplinski committed:
Comments Query: Reformat comment-template.php

gziolo commented on packages/block-library/src/comment-template/index.php:
```diff
@@ -5,12 +5,51 @@
+		'post_id' => $postID
+	);
+
+	$comments = get_comments($args);
```
Nit: you will have to run `npm run format-php` at some point to ensure that PHP code follows coding style guidelines.

gziolo replied:
I see you did that already in the follow-up commits. You can ignore my previous comment. I used a wrong view 🙈

gziolo commented on packages/block-library/src/comment-template/edit.js:
```diff
@@ -0,0 +1,84 @@
+} from '@wordpress/block-editor';
+import { Spinner } from '@wordpress/components';
+import { store as coreStore } from '@wordpress/core-data';
+import { store as editorStore } from '@wordpress/editor';
```
Not a blocker, but we eventually need to find a better way to pass the post id. We can do it later because it's a complex issue that exists also in other places. It isn't easy to explain the motivation behind it. The same issue is tracked in https://github.com/WordPress/gutenberg/issues/32842 for other core blocks.

We have this section in the documentation that explains the differences between several editor packages:
https://github.com/WordPress/gutenberg/blob/trunk/docs/explanations/architecture/modularity.md#whats-the-difference-between-the-different-editor-packages-whats-the-purpose-of-each-package. The gist of it is that we try to have core blocks independent of the existing context and that's why we make them work without `@wordpress/editor` package as a dependency. This way the block will operate also when inserted on different screens like the widget editor screen that knows nothing about the currently edited post id.

ntsekouras replied:
`postId` actually exists in the `defaultBlockContext`: https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/components/provider/index.js#L107

So we just need to add `postId` in the block.json `usesContext`.

michalczaplinski replied:
Aaah, okay! 🙂 I didn't know about this, really useful to know. Thanks both of you!

michalczaplinski replied:
Maybe a dumb question but is there a preferred way to get the post ID in PHP, too? 

Because I could get it using `$block->context['query']` or `get_the_ID()` or probably some other method that I don't know about.

michalczaplinski replied:
Oh, I've just seen [your suggestion below ](https://github.com/WordPress/gutenberg/pull/35231#discussion_r736395005) to use `$block->context['query']` thanks :)

Still, could you explain why this is preferred?

ntsekouras replied:
This way a block can work without depending on the current screen and therefore the `currently edited post id` and whether it exists or not. An example of this is the Post Blocks that are used inside the `Query Loop`. If these blocks used `get_the_ID`, they would all show info from the current post/page - instead we pass them other post ids through the context. Makes sense?

michalczaplinski replied:
Ahh, yea of course! 🙂

gziolo commented on packages/block-library/src/comments-query-loop/toolbar.js:
```diff
@@ -0,0 +1,59 @@
+
+export default function CommentsQueryLoopToolbar( { queryPerPage, setQuery } ) {
+	return (
+		<>
```
Nitpick: the React Fragment wrapper is redundant at the moment.

gziolo commented on packages/block-library/src/comments-query-loop/index.php:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Registers the `core/comments-query-loop` block on the server.
+ */
+function register_block_core_comments_query() {
```
Nitpick: if we don't plan to provide a `render_callback` here we can skip the PHP file and list the `block.json` instead in:

https://github.com/WordPress/gutenberg/blob/2266ff8e22a681ddf4148e10585dc01e11920abd/lib/blocks.php#L15-L16

michalczaplinski replied:
yup, makes sense.

gziolo commented on packages/block-library/package.json:
```diff
@@ -46,6 +46,7 @@
 		"@wordpress/date": "file:../date",
 		"@wordpress/deprecated": "file:../deprecated",
 		"@wordpress/dom": "file:../dom",
+		"@wordpress/editor": "file:../editor",
```
https://github.com/WordPress/gutenberg/runs/4003771134?check_suite_focus=true

<img width="1479" alt="Screen Shot 2021-10-26 at 10 52 23" src="https://user-images.githubusercontent.com/699132/138844714-ca273898-5c6f-4ae7-87f1-a753695cf5e3.png">

`npm install` should resolve the issue with the lock file.

michalczaplinski replied:
I only needed the `@wordpress/editor` in order to get the `postId` which I can now get from the context, so this issue should go away I think 🙂

gziolo reviewed (commented):
Great progress on this PR. Everything is shaping up nicely. We are at the nitpicking stage which is a great sign 😄 

We need now to test whether the current version is good enough to land those two new block in the Gutenberg plugin.

ntsekouras commented on packages/block-library/src/comment-template/block.json:
```diff
@@ -0,0 +1,15 @@
+	"parent": [ "core/comments-query-loop" ],
+	"description": "Contains the block elements used to render a comment, like the title, date, author, avatar and more.",
+	"textdomain": "default",
+	"usesContext": [ "queryId", "queryPerPage" ],
```
```suggestion
	"usesContext": [ "postId", "queryId", "queryPerPage" ],
```

michalczaplinski replied:
Ah, I've already done this change before I've seen your suggestion 👍

ntsekouras commented on packages/block-library/src/comment-template/edit.js:
```diff
@@ -0,0 +1,84 @@
+	const { comments, blocks } = useSelect( ( select ) => {
+		const { getEntityRecords } = select( coreStore );
+		const { getBlocks } = select( blockEditorStore );
+		const postId = select( editorStore ).getCurrentPostId();
```
We should extract this from `context` with [this change](https://github.com/WordPress/gutenberg/pull/35231/files#r736380018)

michalczaplinski replied:
yup, done 🙂

ntsekouras commented on packages/block-library/src/comment-template/index.php:
```diff
@@ -0,0 +1,56 @@
+function render_block_core_comment_template( $attributes, $content, $block ) {
+
+	// Get an array of comments for the current post.
+	$comments = get_approved_comments( get_the_ID() );
```
Related with the context you should extract the id from context: `$block->context['postId']`.

Nit: Also I guess you could bail early if the `postId` isn't set, instead of passing it to the `get_approved_comments`.

ntsekouras commented on packages/block-library/src/comment-template/edit.js:
```diff
@@ -0,0 +1,84 @@
+import { store as coreStore } from '@wordpress/core-data';
+import { store as editorStore } from '@wordpress/editor';
+
+const TEMPLATE = [
```
@kjellr , @jameskoster do you have any suggestions for the default template?

ntsekouras commented on packages/block-library/src/comment-template/edit.js:
```diff
@@ -0,0 +1,84 @@
+		const postId = select( editorStore ).getCurrentPostId();
+
+		return {
+			comments: getEntityRecords( 'root', 'comment', { post: postId } ),
```
You should also use the passed `queryPerPage` from context in this query.

michalczaplinski replied:
yes, thank you! I was gonna do that 😅

ntsekouras commented on packages/block-library/src/comments-query-loop/edit.js:
```diff
@@ -0,0 +1,65 @@
+const TEMPLATE = [ [ 'core/comment-template' ] ];
+
+export default function CommentsQueryLoopEdit( { attributes, setAttributes } ) {
+	const { queryPerPage, tagName: TagName, layout = {} } = attributes;
```
`layout` will never have a value because the `__experimentalLayout` support in block.json is missing. As we had discussed before, I think it's okay to loose the `layout` for the first iteration. 

What this means is that we don't need all the `layout` retrievals/checks in this file and we will not pass the prop in `useInnerBlocksProps`. This will just result in a block which will get all the available space from it's parent container - which is just fine for now.

michalczaplinski replied:
yup, gotcha - forgot to remove this earlier.

ntsekouras reviewed (commented):
Great work here @michalczaplinski ! This is getting in very good shape! 💯 

Another small thing is that we will need a small styling fix for `ul` to not display the bullet [like here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/post-template/style.scss#L1).

michalczaplinski committed:
Comments Query: Use the postId from the block context

michalczaplinski committed:
Comments Query: Remove redundant Fragment in toolbar

michalczaplinski committed:
Comments Query: Remove unnecessary layout stuff

michalczaplinski committed:
Comments Query: Remove .php file and list block.json

michalczaplinski committed:
Comments Query: Get postId from context on server

michalczaplinski committed:
Comments Query: Pass queryPerPage to getEntityRecords

michalczaplinski committed:
Comments Query: Pass queryPerPage to get_approved_comments

michalczaplinski commented:
Thanks a lot for your feedback @gziolo & @ntsekouras 

I've pushed some more updates but I will still have some small changes to make (non-exhaustive list):

- [x] ~I think I'd rename `queryPerPage` to `numberOfComments`. While the argument to getEntityRecords is called `per_page`, the argument to the [`get_comments()` PHP function](https://developer.wordpress.org/reference/functions/get_comments/) is just called `number`. While `numberOfComments` is on the verbose side, I like that it's very clear.~
- [x] ~I think there are still some e2e tests failures.~
- [x] Make sure that we specify the default `order` of comments because it seems to be different when using `getEntityRecords( 'root', 'comment', query)` in JS and `get_approved_comments()` in PHP.
- [ ] Possibly inherit the `commentsNumber` from the discussion settings.
- [ ] Make sure we handle when comments are disabled / closed. 
- [ ] Handle nested comments gracefully (not sure how yet?)
- [ ] Styling those `<ul>` elements
- [ ] Use a comment query instead of `get_approved_comments()` ?
- [ ] I've noticed that there is some FOUC in the block when I load the page - I want to see why
- [ ] Make sure the comment-template icon is correct: https://github.com/WordPress/gutenberg/pull/35231#discussion_r731014677
- [ ] Confirm which blocks to use in the default template: https://github.com/WordPress/gutenberg/pull/35231#discussion_r736396805

carolinan commented on packages/block-library/src/comment-template/index.php:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	$content = '';
+	foreach ( $comments as $comment ) {
```
I don't believe that the pagination will work unless a comment query is used.

michalczaplinski replied:
Could you explain why it wouldn't work? I see that [`get_approved_comments()`](https://developer.wordpress.org/reference/functions/get_approved_comments/#source) uses the comments query under the hood.

ntsekouras replied:
Yeah, it seems it's just a small wrapper of `WP_Comment_Query`, so it should be fine. We will find out for sure when the pagination blocks will be implemented 😄

gziolo commented:
>  I think I'd rename queryPerPage to numberOfComments. While the argument to getEntityRecords is called per_page, the argument to the get_comments() PHP function is just called number. While numberOfComments is on the verbose side, I like that it's very clear.

Definitely, `number ` set as an attribute alone might be confusing. There is an `offset` already, so something related to the limit (like in SQL) could also work. Naming is hard, for some folks number of comments might be read as the total number of comments. I know I'm not helping 🤷🏻 😅 

In general, I think that all functional changes you mentioned can be done in follow-up PRs so the Gutenberg plugin users can start experimenting with how all newly created blocks work together.

gziolo commented:
The good news is that this PR will also fix https://github.com/WordPress/gutenberg/issues/35000 as far as I can see by looking at the requirements 🎉 I reflected that in the description so GitHub can do its magic when this PR lands.

michalczaplinski committed:
Fix block names array that I messed up while rebasing

michalczaplinski commented:
> Definitely, number set as an attribute alone might be confusing. There is an offset already, so something related to the limit (like in SQL) could also work. Naming is hard, for some folks number of comments might be read as the total number of comments. I know I'm not helping 🤷🏻 😅

Yeah, I was also thinking of `limit`. Naming is hard indeed. I've realized that we will add pagination later on so then `queryPerPage` makes more sense. Will just keep it as is for now.

michalczaplinski commented:
My current concern is how to handle the nested comments.

It seems to me that that the best way would be to convert the list of comments passed in `blockContexts`: https://github.com/WordPress/gutenberg/blob/1668cd38e9718acba32ba5a4df2b597a5706fb22/packages/block-library/src/comment-template/edit.js#L47-L52

to a format that includes an array of `children`. This is possible because each comment object also stores an ID of its parent comment.  


```js
[
	{ commentId: 1, children: [] },
	{
		commentId: 2, children: [
			{ commentId: 3, children: [] },
			{ commentId: 4, children: [ 
			   { commentId: 5, children: [] } 
		  ] },
		],
	},
	{ commentId: 3, children: [] },
];
```

You get the idea.

Probably would need to split the `<CommentTemplateEdit/> into 2 components. The first one that contains the logic and the second one that "just" renders the content given some props. 

Then, that second component call itself recursively (if the comment has children) so that I can represent each level of comment nesting.

ntsekouras commented:
>My current concern is how to handle the nested comments.

I think the first iteration could skip this and after landing, create a quick follow up for the nested handling. What do you think?

ntsekouras commented on packages/block-library/src/comment-template/edit.js:
```diff
@@ -0,0 +1,91 @@
+			const { getBlocks } = select( blockEditorStore );
+
+			return {
+				comments: getEntityRecords( 'root', 'comment', {
```
You will need to add the same default arguments used in `get_approved_comments`
```
status: 1,
order: 'asc',
```

gziolo replied:
Those are default values according to https://developer.wordpress.org/rest-api/reference/comments/#arguments, but we can make it explicit 👍🏻

ntsekouras replied:
The default order was 'desc' - that was the missing piece 😄

gziolo commented on packages/block-library/src/comment-template/edit.js:
```diff
@@ -0,0 +1,91 @@
+			return {
+				comments: getEntityRecords( 'root', 'comment', {
+					post: postId,
+					per_page: queryPerPage,
```
Based on https://developer.wordpress.org/rest-api/reference/comments/#arguments those are default values, but it's better to be explicit here:

```suggestion
					status: 'approve',
					per_page: queryPerPage,
					order: 'asc',
```

gziolo committed:
Update packages/block-library/src/comment-template/edit.js

gziolo reviewed (approved):
Awesome work. I think it's in good shape already so we can land it to start testing in the Gutenberg plugin. We still have almost a week before the 11.9 RC1 happens to implement more features necessary for the Comments Query Loop block. The biggest benefit of landing this PR early is that we can start deprecating the Post Comment block in favor of the Comment Template block.

ntsekouras commented:
Great work! 🚀
